nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java

```diff
@@ -0,0 +1,537 @@
+import static org.apache.nifi.atlas.NiFiTypes.ATTR_NAME;
+import static org.apache.nifi.atlas.NiFiTypes.ATTR_OUTPUTS;
+import static org.apache.nifi.atlas.NiFiTypes.ATTR_OUTPUT_PORTS;
+import static org.apache.nifi.atlas.NiFiTypes.ATTR_QUALIFIED_NAME;
+import static org.apache.nifi.atlas.NiFiTypes.ATTR_QUEUES;
+import static org.apache.nifi.atlas.NiFiTypes.ATTR_TYPENAME;
+import static org.apache.nifi.atlas.NiFiTypes.ATTR_URL;
+import static org.apache.nifi.atlas.NiFiTypes.ENTITIES;
+import static org.apache.nifi.atlas.NiFiTypes.TYPE_NIFI_FLOW;
+import static org.apache.nifi.atlas.NiFiTypes.TYPE_NIFI_FLOW_PATH;
+import static org.apache.nifi.atlas.NiFiTypes.TYPE_NIFI_INPUT_PORT;
+import static org.apache.nifi.atlas.NiFiTypes.TYPE_NIFI_OUTPUT_PORT;
+import static org.apache.nifi.atlas.NiFiTypes.TYPE_NIFI_QUEUE;
+
+public class NiFiAtlasClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(NiFiAtlasClient.class);
+
+    private static NiFiAtlasClient nifiClient;
+    private AtlasClientV2 atlasClient;
+
+    private NiFiAtlasClient() {
+        super();
+    }
+
+    public static NiFiAtlasClient getInstance() {
+        if (nifiClient == null) {
+            synchronized (NiFiAtlasClient.class) {
+                if (nifiClient == null) {
+                    nifiClient = new NiFiAtlasClient();
+                }
+            }
+        }
+        return nifiClient;
+    }
+
+    public void initialize(final String[] baseUrls, final AtlasAuthN authN, final File atlasConfDir) {
+
+        synchronized (NiFiAtlasClient.class) {
+
+            if (atlasClient != null) {
+                logger.info("{} had been setup but replacing it with new one.", atlasClient);
+                ApplicationProperties.forceReload();
+            }
+
+            if (atlasConfDir != null) {
+                // If atlasConfDir is not set, atlas-application.properties will be searched under classpath.
+                Properties props = System.getProperties();
+                final String atlasConfProp = "atlas.conf";
+                props.setProperty(atlasConfProp, atlasConfDir.getAbsolutePath());
+                logger.debug("{} has been set to: {}", atlasConfProp, props.getProperty(atlasConfProp));
+            }
+
+            atlasClient = authN.createClient(baseUrls);
+
+        }
+    }
+
+    /**
+     * This is an utility method to delete unused types.
+     * Should be used during development or testing only.
+     * @param typeNames to delete
+     */
+    void deleteTypeDefs(String ... typeNames) throws AtlasServiceException {
+        final AtlasTypesDef existingTypeDef = getTypeDefs(typeNames);
+        try {
+            atlasClient.deleteAtlasTypeDefs(existingTypeDef);
+        } catch (UniformInterfaceException e) {
+            if (e.getResponse().getStatus() == 204) {
+                // 204 is a successful response.
+                // NOTE: However after executing this, Atlas should be restarted to work properly.
```

Comment:
Is this a bug in Atlas? If so, has a JIRA been filed for the Atlas team? Should we be logging a warning here?

Reply:
I reported several findings to Atlas project while I work on this integration. However, I think I haven't share this particular issue yet. Thanks, I will do. This method is only used by ITTest class, though.